libtenzir/builtins/formats/yaml.cpp

```diff
@@ -0,0 +1,286 @@
+//   |___/_/ |_/___/ /_/       Space and Time
+//
+// SPDX-FileCopyrightText: (c) 2023 The Tenzir Contributors
+// SPDX-License-Identifier: BSD-3-Clause
+
+#include <tenzir/adaptive_table_slice_builder.hpp>
+#include <tenzir/argument_parser.hpp>
+#include <tenzir/arrow_table_slice.hpp>
+#include <tenzir/concept/parseable/tenzir/data.hpp>
+#include <tenzir/concept/parseable/to.hpp>
+#include <tenzir/data.hpp>
+#include <tenzir/error.hpp>
+#include <tenzir/plugin.hpp>
+#include <tenzir/table_slice.hpp>
+#include <tenzir/to_lines.hpp>
+#include <tenzir/type.hpp>
+#include <tenzir/view.hpp>
+
+#include <arrow/record_batch.h>
+#include <caf/error.hpp>
+#include <caf/expected.hpp>
+#include <fmt/format.h>
+
+#include <yaml-cpp/yaml.h>
+
+namespace tenzir::plugins::yaml {
+namespace {
+
+constexpr auto document_end_marker = "...";
+constexpr auto document_start_marker = "---";
+
+auto parse_node(auto& guard, const YAML::Node& node,
+                operator_control_plane& ctrl) -> void {
+  switch (node.Type()) {
+    case YAML::NodeType::Undefined:
+    case YAML::NodeType::Null: {
+      break;
+    }
+    case YAML::NodeType::Scalar: {
+      if (auto as_bool = bool{}; YAML::convert<bool>::decode(node, as_bool)) {
+        if (auto err = guard.add(as_bool)) [[unlikely]] {
+          diagnostic::error("failed to append value: {}", err)
+            .emit(ctrl.diagnostics());
+        }
+        break;
+      }
+      // TODO: Do not attempt to parse pattern, map, list, and record here.
+      if (auto as_data = data{}; parsers::data(node.Scalar(), as_data)) {
+        if (auto err = guard.add(make_data_view(as_data))) [[unlikely]] {
+          diagnostic::error("failed to append value: {}", err)
+            .emit(ctrl.diagnostics());
+        }
+        break;
+      }
+      if (auto err = guard.add(make_view(node.Scalar()))) [[unlikely]] {
+        diagnostic::error("failed to append value: {}", err)
+          .emit(ctrl.diagnostics());
+      }
+      break;
+    }
+    case YAML::NodeType::Sequence: {
+      auto list = guard.push_list();
+      for (const auto& element : node) {
+        parse_node(list, element, ctrl);
+      }
+      break;
+    }
+    case YAML::NodeType::Map: {
+      auto record = guard.push_record();
+      for (const auto& element : node) {
+        auto field = record.push_field(element.first.Scalar());
+        parse_node(field, element.second, ctrl);
+      }
+      break;
+    }
+  }
+};
+
+auto load_document(adaptive_table_slice_builder& builder,
+                   std::string&& document, operator_control_plane& ctrl)
+  -> void {
+  auto row = builder.push_row();
+  try {
+    auto node = YAML::Load(document);
+    if (not node.IsMap()) {
+      diagnostic::error("document is not a map").emit(ctrl.diagnostics());
+      return;
+    }
+    for (const auto& element : node) {
+      auto field = row.push_field(element.first.as<std::string>());
+      parse_node(field, element.second, ctrl);
```

Comment:
There's a stack-use-after-scope somewhere in the adaptive builder here that this code reveals; since @jachris is working on fixing that for other reasons anyways, I think it makes sense to wait with this PR until after that has been fixed.